controllers/virtinkmachine_controller_test.go

```diff
@@ -0,0 +1,192 @@
+						Instance: virtv1alpha1.Instance{
+							CPU: virtv1alpha1.CPU{
+								Sockets:        uint32(1),
+								CoresPerSocket: uint32(2),
+							},
+						},
+					},
+				},
+			}
+			Expect(controllerutil.SetOwnerReference(&machine, &virtinkMachine, k8sClient.Scheme())).To(Succeed())
+			Expect(k8sClient.Create(ctx, &virtinkMachine)).To(Succeed())
+		})
+
+		It("should add finalizer", func() {
+			var virtinkMachine infrastructurev1beta1.VirtinkMachine
+			Eventually(func() bool {
+				Expect(k8sClient.Get(ctx, virtinkMachineKey, &virtinkMachine)).To(Succeed())
+				return controllerutil.ContainsFinalizer(&virtinkMachine, finalizer)
+			}).Should(BeTrue())
+		})
+
+		Context("when infrastructure cluster is not ready", func() {
+			It("should not create VM", func() {
+				var vm virtv1alpha1.VirtualMachine
+				Consistently(func() bool {
+					return apierrors.IsNotFound(k8sClient.Get(ctx, virtinkMachineKey, &vm))
+				}).Should(BeTrue())
+			})
+		})
+
+		Context("when infrastructure cluster is ready", func() {
+			BeforeEach(func() {
+				var cluster capiv1beta1.Cluster
+				Expect(k8sClient.Get(ctx, clusterKey, &cluster)).To(Succeed())
+				cluster.Status.InfrastructureReady = true
+				Expect(k8sClient.Status().Update(ctx, &cluster)).To(Succeed())
+			})
+
+			Context("when bootstrap data secret is not set", func() {
+				It("should not create VM", func() {
+					var vm virtv1alpha1.VirtualMachine
+					Consistently(func() bool {
+						return apierrors.IsNotFound(k8sClient.Get(ctx, virtinkMachineKey, &vm))
+					}).Should(BeTrue())
+				})
+			})
+
+			Context("when bootstrap data secret is set", func() {
+				BeforeEach(func() {
+					var machine capiv1beta1.Machine
+					Expect(k8sClient.Get(ctx, machineKey, &machine)).To(Succeed())
+					secretName := machine.Name + "-" + "secret"
+					secret := corev1.Secret{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      secretName,
+							Namespace: machine.Namespace,
+						},
+						StringData: map[string]string{
+							"value": "#cloud-init",
+						},
+					}
+					Expect(k8sClient.Create(ctx, &secret)).To(Succeed())
+
+					machine.Spec.Bootstrap.DataSecretName = &secretName
+					Expect(k8sClient.Update(ctx, &machine)).To(Succeed())
+				})
+
+				It("should create a virtink VM", func() {
+					var vm virtv1alpha1.VirtualMachine
+					Eventually(func() bool {
+						return k8sClient.Get(ctx, virtinkMachineKey, &vm) == nil
+					}).Should(BeTrue())
+					Expect(vm.Spec.Volumes[0].CloudInit.UserDataBase64).To(Equal(base64.StdEncoding.EncodeToString([]byte("#cloud-init"))))
+
+					var virtinkMachine infrastructurev1beta1.VirtinkMachine
+					Eventually(func() bool {
+						Expect(k8sClient.Get(ctx, machineKey, &virtinkMachine)).To(Succeed())
+						return *virtinkMachine.Spec.ProviderID == fmt.Sprintf("virtink://%s", vm.UID) &&
+							virtinkMachine.Status.Ready
+					})
+				})
+
+				Context("when deleting VirtinkMachine", func() {
```

Comment:
Maybe context "deleting VirtinkMachine" should not be contained by context "for a pending VirtinkMachine" ?

Reply:
"deleting VirtinkMachine"  is not directly in "for a pending VirtinkMachine" context,  there are multilayer context. this way could use some public parent context to reduce repetitive  codes.